### PR TITLE
Refactor feature gates

### PR DESCRIPTION
**Pull request by killercup**
After #63824, this goes a few steps further by

- parsing doc comments in the macros to extract descriptions for feature gates, and
- introducing a common `Feature` type to replace the tuples used previously to improve readability.

The descriptions are not yet used, but I felt like this PR is a useful enough refactoring on its own.

r? @Centril

### CODE REVIEW

File: src/libsyntax/feature_gate/check.rs [hunk 1]
```diff
@@ -733,10 +732,14 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
         }
     }
 
-    for &(name, .., f_edition, set) in ACTIVE_FEATURES {
-        if let Some(f_edition) = f_edition {
-            if f_edition <= crate_edition {
-                set(&mut features, DUMMY_SP);
+    for &Feature { name, edition: f_edition, state, .. } in ACTIVE_FEATURES {
+        if let Some(edition) = f_edition {
+            if edition <= crate_edition {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]: This `if let .. { if ... {` could be turned into `ACTIVE_FEATURES.filter(...)`

File: src/libsyntax/feature_gate/check.rs [hunk 2]
```diff
@@ -766,12 +769,16 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
                     continue;
                 }
 
-                for &(name, .., f_edition, set) in ACTIVE_FEATURES {
+                for &Feature { name, edition: f_edition, state, .. } in ACTIVE_FEATURES {
                     if let Some(f_edition) = f_edition {
                         if f_edition <= *edition {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]: ...Same here, except `crate_edition` substituted for `edition`. Could use a function for these, e.g. `fn active_features_until_edition(edition: ...)`.

File: src/libsyntax/feature_gate/check.rs [hunk 3]
```diff
@@ -733,10 +732,14 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
...
     }
 
-    for &(name, .., f_edition, set) in ACTIVE_FEATURES {
-        if let Some(f_edition) = f_edition {
-            if f_edition <= crate_edition {
-                set(&mut features, DUMMY_SP);
+    for &Feature { name, edition: f_edition, state, .. } in ACTIVE_FEATURES {
+        if let Some(edition) = f_edition {
+            if edition <= crate_edition {
+                match state {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]: This...

File: src/libsyntax/feature_gate/check.rs [hunk 4]
```diff
@@ -766,12 +769,16 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
...
                 }
 
-                for &(name, .., f_edition, set) in ACTIVE_FEATURES {
+                for &Feature { name, edition: f_edition, state, .. } in ACTIVE_FEATURES {
                     if let Some(f_edition) = f_edition {
                         if f_edition <= *edition {
                             // FIXME(Manishearth) there is currently no way to set
                             // lib features by edition
-                            set(&mut features, DUMMY_SP);
+                            match state {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]: ...is repeated here and could be refactored?

File: src/libsyntax/feature_gate/check.rs [hunk 5]
```diff
@@ -829,14 +836,21 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
...
 
-            let removed = REMOVED_FEATURES.iter().find(|f| name == f.0);
-            let stable_removed = STABLE_REMOVED_FEATURES.iter().find(|f| name == f.0);
-            if let Some((.., reason)) = removed.or(stable_removed) {
-                feature_removed(span_handler, mi.span(), *reason);
-                continue;
+            let removed = REMOVED_FEATURES.iter().find(|f| name == f.name);
+            let stable_removed = STABLE_REMOVED_FEATURES.iter().find(|f| name == f.name);
+            if let Some(Feature { state, .. }) = removed.or(stable_removed) {
+                match state {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]:
```suggestion
                if let FeatureState::Removed { reason } | FeatureState::Stabilized { reason } = state {
```

File: src/libsyntax/feature_gate/check.rs [hunk 6]
```diff
@@ -851,8 +865,12 @@ pub fn get_features(span_handler: &Handler, krate_attrs: &[ast::Attribute],
                 }
             }
 
-            if let Some((.., set)) = ACTIVE_FEATURES.iter().find(|f| name == f.0) {
-                set(&mut features, mi.span());
+            if let Some(Feature { state, .. }) = ACTIVE_FEATURES.iter().find(|f| name == f.name) {
+                match state {
```
Comment (Centril) [inline src/libsyntax/feature_gate/check.rs]: ...repeated 3rd time here.

**Review by Centril — CHANGES_REQUESTED** [review]
Thanks! r=me with comments below addressed. :)

**Comment by killercup** [comment]
Thanks for the review, @Centril! I've changed most of the loops you mentioned, so it'd be good to have a second look before I blindly r+ this :)

**Review by Centril — APPROVED** [review]

**Comment by Centril** [comment]
That's odd; I don't see anything that would cause this... @bors r-

**Comment by killercup** [comment]
Let's try this again! Tests pass locally but I'll wait for CI before sending it off to bors.

**Pull request merged**